src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,244 @@
+		}
+
+		public void Handle(RedactionMessage.GetEventPosition message) {
+			ThreadPool.QueueUserWorkItem(_ => {
+				try {
+					GetEventPosition(message.EventStreamId, message.EventNumber, message.Envelope);
+				} catch (Exception ex) {
+					Log.Error(ex, "An error has occurred when getting position for stream: {stream}, event number: {eventNumber}.",
+						message.EventStreamId, message.EventNumber);
+					message.Envelope.ReplyWith(
+						new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.UnexpectedError, Array.Empty<EventPosition>()));
+				}
+			});
+		}
+
+		private void GetEventPosition(string streamName, long eventNumber, IEnvelope envelope) {
+			var streamId = _readIndex.GetStreamId(streamName);
+			var result = _readIndex.ReadEventInfo_KeepDuplicates(streamId, eventNumber);
+
+			var eventPositions = new EventPosition[result.EventInfos.Length];
+
+			for (int i = 0; i < result.EventInfos.Length; i++) {
+				var eventInfo = result.EventInfos[i];
+				var logPos = eventInfo.LogPosition;
+				var chunk = _db.Manager.GetChunkFor(logPos);
+				var localPosition = chunk.ChunkHeader.GetLocalLogPosition(logPos);
+				var chunkEventOffset = chunk.GetActualRawPosition(localPosition);
+
+				// all the events returned by ReadEventInfo_KeepDuplicates() must exist in the log
+				// since the log record was read from the chunk to check for hash collisions.
+				if (chunkEventOffset < 0)
+					throw new Exception($"Failed to fetch actual raw position for event at log position: {logPos}");
+
+				if (chunkEventOffset > uint.MaxValue)
+					throw new Exception($"Actual raw position for event at log position: {logPos} is larger than uint.MaxValue: {chunkEventOffset}");
+
+				eventPositions[i] = new EventPosition(
+					logPosition: logPos,
+					chunkFile: Path.GetFileName(chunk.FileName),
+					chunkVersion: chunk.ChunkHeader.Version,
+					chunkComplete: chunk.ChunkFooter is { IsCompleted: true },
+					chunkEventOffset: (uint) chunkEventOffset);
+			}
+
+			envelope.ReplyWith(
+				new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.Success, eventPositions));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkLock message) {
+			if (_switchChunksLock.TryAcquire())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkUnlock message) {
+			if (_switchChunksLock.TryRelease())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunk message) {
+			ThreadPool.QueueUserWorkItem(_ => {
+				try {
```

Comment:
if we had the lock at the time this was scheduled, is it possible that we released it before this actually runs

im not certain, but it might be sensible to give the redaction service its own queue (channel?) similar to the writerservice so that it can just be conceptually single threaded and process off of the main bus. then the lock/switch/unlock steps would be performed strictly one at a time

Reply:
i'm not sure i followed, do you mean:
- lock acquired
- chunk switch has started (but envelope not yet replied)
- lock released due to gRPC heartbeat timeout (e.g if the redactor app exits)
- chunk switch completes, replying on the envelope

in this case, it seems that it wouldn't cause an issue on the gRPC side, since the tcs would be set but nothing would be awaiting it any more

Reply:
fixed in: 56fadc4f319ef10d126217f138452cb3e4d809ef. for some reason, the queue stats don't seem to work, i'm looking into it.